src/tests/p11test/p11test_case_common.c

```diff
@@ -585,39 +575,45 @@ int callback_public_keys(test_certs_t *objects,
 			if (EVP_PKEY_get_group_name(o->key, curve_name, sizeof(curve_name), &curve_name_len)||
 				EVP_PKEY_get_octet_string_param(o->key, OSSL_PKEY_PARAM_PUB_KEY, pubkey, sizeof(pubkey), &pubkey_len) != 1) {
 				debug_print(" [WARN %s ] Can not get params from EVP_PKEY", o->id_str);
-				EC_GROUP_free(ecgroup);
-				EC_POINT_free(ecpoint);
-				return -1;
+				goto ec_error;
 			}
 
 			if (!(bld = OSSL_PARAM_BLD_new()) ||
 				OSSL_PARAM_BLD_push_utf8_string(bld, "group", curve_name, curve_name_len) != 1 ||
 				OSSL_PARAM_BLD_push_octet_string(bld, "pub", pubkey, pubkey_len) != 1 ||
 				!(params = OSSL_PARAM_BLD_to_param(bld))) {
 				debug_print(" [WARN %s ] Can not set params from EVP_PKEY", o->id_str);
-				EC_GROUP_free(ecgroup);
-				EC_POINT_free(ecpoint);
-				OSSL_PARAM_BLD_free(bld);
-				EVP_PKEY_CTX_free(ctx);
-				return -1;
+				goto ec_error;
 			}
-			OSSL_PARAM_BLD_free(bld);
 
 			if (ctx == NULL || params == NULL ||
 				EVP_PKEY_fromdata_init(ctx) != 1 ||
 				EVP_PKEY_fromdata(ctx, &o->key, EVP_PKEY_PUBLIC_KEY, params) != 1) {
 				debug_print(" [WARN %s ] Can not set params for EVP_PKEY", o->id_str);
-				EC_GROUP_free(ecgroup);
-				EC_POINT_free(ecpoint);
-				EVP_PKEY_CTX_free(ctx);
-				return -1;
+				goto ec_error;
 			}
-			EVP_PKEY_CTX_free(ctx);
-			OSSL_PARAM_free(params);
 #endif
 		}
+
+		int ec_error = 0;
+		goto ec_out;
+
+	ec_error:
+		ec_error = 1;
+
+	ec_out:
 		EC_GROUP_free(ecgroup);
 		EC_POINT_free(ecpoint);
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+		EVP_PKEY_CTX_free(ctx);
+		OSSL_PARAM_BLD_free(bld);
+		OSSL_PARAM_free(params);
+		EC_GROUP_free(cert_group);
+		EC_POINT_free(cert_point);
+#endif
+
+		if (ec_error) return -1;
+
```

Comment:
please move the `int ec_error` definition to the beginning of the block. I was thinking how to make this less jumping as we are not much happy to use `goto`/labels in the new code. What would you say to the following suggestion that would use only one label?
```
		int ec_error = 1; /* initially */

		/* the code above with the `goto ec_out` jumps */

		ec_error = 0;

ec_out:
		EC_GROUP_free(ecgroup);
		EC_POINT_free(ecpoint);
#if OPENSSL_VERSION_NUMBER >= 0x30000000L
		EVP_PKEY_CTX_free(ctx);
		OSSL_PARAM_BLD_free(bld);
		OSSL_PARAM_free(params);
		EC_GROUP_free(cert_group);
		EC_POINT_free(cert_point);
#endif

		if (ec_error)
			return -1;

```

Reply:
Excellent suggestion applied.